### PR TITLE
mimxrt/adc: Added ADC support.

### PR DESCRIPTION
**Pull request by alphaFred**
Added support for ADC module. Thanks again for the great support from @robert-hh.

### CODE REVIEW

**Comment by alphaFred** [comment]
Hi @dpgeorge 
I have a new addition to the mimxrt port ready. Looking forward to your review.

Thanks to the always great support of @robert-hh I can even provide you with a picture of a 0.0 - 3.0V ramp :smile: 
image

With x-axis being sample number and y-axis being raw value of `read_u16()`.

File: ports/mimxrt/Makefile [hunk 1]
```diff
@@ -101,12 +101,15 @@ SRC_TINYUSB_IMX_C += \
...
 SRC_C = \
 	main.c \
+	adc.c \
 	led.c \
 	pin.c \
 	ticks.c \
 	tusb_port.c \
 	board_init.c \
 	$(BOARD_DIR)/flash_config.c \
+	$(MCU_DIR)/drivers/fsl_adc.c \
```
Comment (dpgeorge) [inline ports/mimxrt/Makefile:111]: please place this line in the `SRC_TINYUSB_IMX_C` macro above (which is wrongly named, but at least it's the right place)

File: ports/mimxrt/adc.c [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        // Configure ADC perpheral
+        adc_config_t config;
+        ADC_GetDefaultConfig(&config);
+        ADC_Init(adc_instance, &config);
+
+        // Perform calibration
+        status_t calib_state = ADC_DoAutoCalibration(adc_instance);
+
+        if (calib_state == kStatus_Fail) {
+            nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_ValueError, "Calibration for ADC Instance %d failed", i));
```
Comment (dpgeorge) [inline ports/mimxrt/adc.c:49]: This cannot raise an exception at this point, there is nothing to catch it.  Instead just use `mp_printf(&mp_plat_print, ...)`.

File: ports/mimxrt/adc.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+typedef struct _machine_adc_obj_t {
+    mp_obj_base_t base;
+    ADC_Type *adc;
+    uint8_t channel;
+    uint8_t channel_group;
+    uint16_t resolution;
+} machine_adc_obj_t;
+
+extern const mp_obj_type_t machine_adc_type;
```
Comment (dpgeorge) [inline ports/mimxrt/adc.h:43]: This type and the typedef just above it really belong in `modmachine.h`.

File: ports/mimxrt/machine_adc.c [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+#include "py/runtime.h"
+#include "py/mphal.h"
+
+#include "fsl_gpio.h"
+#include "fsl_iomuxc.h"
+#include "adc.h"
+
+
+STATIC ADC_Type *const adcBases[] = ADC_BASE_PTRS;
```
Comment (dpgeorge) [inline ports/mimxrt/machine_adc.c:35]: please don't use camel case, rather call it `adc_bases`

File: ports/mimxrt/adc.c [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ * THE SOFTWARE.
+ */
+
+#include "py/runtime.h"
+#include "py/mphal.h"
+
+#include "adc.h"
+
+
+void adc_init(void) {
```
Comment (dpgeorge) [inline ports/mimxrt/adc.c:33]: maybe this function can go in `machine_adc.c` just to keep all the code self-contained?

**Review by dpgeorge — CHANGES_REQUESTED** [review]
Thanks for this!  Please rebase on latest master (if possible, otherwise I can do that during merge).

**Comment by alphaFred** [comment]
Hi @dpgeorge 
I will rebase and fix your findings today. Thanks for the review.

File: ports/mimxrt/Makefile (hunk 1 above)
Comment (alphaFred) [reply ports/mimxrt/Makefile:111]: I changed the name of the macro to `SRC_HAL_IMX_C` and added the respective line.

File: ports/mimxrt/adc.h (hunk 3 above)
Comment (alphaFred) [reply ports/mimxrt/adc.h:43]: For now all of the machine modules have that structure in this port. Would it be fine with you if I keep the structure for now but refactor all of the modules (led, pin, adc) in a separate PR?

File: ports/mimxrt/machine_adc.c (hunk 4 above)
Comment (alphaFred) [reply ports/mimxrt/machine_adc.c:35]: Fixed

File: ports/mimxrt/adc.h (hunk 3 above)
Comment (dpgeorge) [reply ports/mimxrt/adc.h:43]: The only existing header file is `modmachine.h`, which contains the types for timer and rtc.  So would be good to put at least `machine_adc_type` in `modmachine.h`.  The rest can be changed later.

File: ports/mimxrt/adc.c (hunk 5 above)
Comment (alphaFred) [reply ports/mimxrt/adc.c:33]: :thinking: should this function then be declared extern in `main.c` or rather create a `machine_adc.h`?

File: ports/mimxrt/adc.h (hunk 3 above)
Comment (alphaFred) [reply ports/mimxrt/adc.h:43]: Ok sure - I can do that

File: ports/mimxrt/adc.c (hunk 2 above)
Comment (alphaFred) [reply ports/mimxrt/adc.c:49]: fixed

File: ports/mimxrt/adc.c (hunk 5 above)
Comment (alphaFred) [reply ports/mimxrt/adc.c]: Forget it - I just noticed how it was done by @robert-hh with the timer module. Fixed

**Comment by alphaFred** [comment]
@dpgeorge 
I implemented all requested changes implemented and rebased the branch :-)

**Comment by dpgeorge** [comment]
Thanks for updating.  Rebased, squashed and merged in b8c65b174f95c1fa1117d4f307a1c5a450607e33 with some minor code formatting changes (removed double blank lines, put `{` at end of line of for statement).

**Pull request closed without merging**